numpy/fft/_pocketfft_umath.cpp

```diff
@@ -0,0 +1,437 @@
+/*
+ * This file is part of pocketfft.
+ * Licensed under a 3-clause BSD style license - see LICENSE.md
+ */
+
+/*
+ *  Main implementation file.
+ *
+ *  Copyright (C) 2004-2018 Max-Planck-Society
+ *  \author Martin Reinecke
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#define PY_SSIZE_T_CLEAN
+#include <assert.h>
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/ufuncobject.h"
+
+#include "npy_config.h"
+
+#define POCKETFFT_NO_MULTITHREADING
+#include "pocketfft/pocketfft_hdronly.h"
+
+/* super-simple catch-all for exceptions */
+#define TRY_FOR_CPP_EXCEPTIONS try {
+#define CATCH_CPP_EXCEPTIONS }                         \
+        catch (...) {                                  \
+            _cpp_exception(std::current_exception());  \
+            return;                                    \
+        }
+
+static void
+_cpp_exception(std::exception_ptr e) {
+    NPY_ALLOW_C_API_DEF
+    NPY_ALLOW_C_API;
+    try {
+        std::rethrow_exception(e);
+    }
+    catch (const std::bad_alloc& e) {
+        PyErr_NoMemory();
+    }
+    catch (const std::exception& e) {
+        PyErr_SetString(PyExc_RuntimeError, e.what());
+    }
+    NPY_DISABLE_C_API;
+}
```

Comment:
I don't really love the way it puts a bracket in the macro (yes other places do too).  This would be an alternative.  Haven't made up my mind if I much prefer it.  I prefer the pattern, but it adds churn further down.
```
/*
 * In order to ensure that C++ exceptions are converted to Python
 * ones before crossing over to the C machinery, we must catch them.
 * This template can be used to wrap a C++ written ufunc to do this via:
 *      wrap_legacy_cpp_ufunc<cpp_ufunc>
 */
template<PyUFuncGenericFunction cpp_ufunc>
static void
wrap_legacy_cpp_ufunc(<full ufunc signature>)
{
    NPY_ALLOW_C_API_DEF
    try {
        cpp_ufunc(<full ufunc signature>)
    }
    catch (std::bad_alloc& e) {
        NPY_ALLOW_C_API;
        PyErr_NoMemory();
        NPY_DISABLE_C_API;
    }
    catch (const std::exception& e) {
        NPY_ALLOW_C_API;
        PyErr_SetString(PyExc_RuntimeError, e.what());
        NPY_DISABLE_C_API;
    }
}
```
Then use `wrap_legacy_cpp_ufunc<irfft_loop<double>>` below.

Reply:
That's a really nice idea -- implemented!